ldms/src/sampler/cxi_sampler/cxi_sampler.c

```diff
@@ -0,0 +1,754 @@
+	tel_rec = ldms_record_create("tel_record");
+	cxi->iface_mid = ldms_record_metric_add(tel_rec,
+						"iface_name", "",
+						LDMS_V_CHAR_ARRAY, 32);
+
+	for (i = 0; i < cxi->tel_files[0].count; i++) {
+		rc = ldms_record_metric_add(tel_rec, cxi->tel_files[0].names[i],
+					    "", LDMS_V_D64, 0);
+		if (rc < 0) {
+			ovis_log(cxi->log, OVIS_LERROR,
+				 "Error %d creating metric '%s'\n",
+				 -rc, cxi->tel_files[0].names[i]);
+			rc = errno;
+			goto err;
+		}
+		/* Cache the record metric index in tel_files */
+		for (j = 0; j < cxi->iface_count; j++) {
+			cxi->tel_files[j].midx[i] = rc;
+		}
+	}
+	tel_rec_mid = ldms_schema_record_add(cxi->schema, tel_rec);
+
+	/* Per interface retry handler record */
+	rh_rec = ldms_record_create("rh_record");
+	rc = ldms_record_metric_add(rh_rec,
+				    "iface_name", "",
+				    LDMS_V_CHAR_ARRAY, 32);
+
+	for (i = 0; i < cxi->rh_files[0].count; i++) {
+		rc = ldms_record_metric_add(rh_rec, cxi->rh_files[0].names[i],
+					    "", LDMS_V_D64, 0);
+		if (rc < 0) {
+			rc = errno;
+			goto err;
+		}
+		/* Cache the record metric index in rh_files */
+		for (j = 0; j < cxi->iface_count; j++) {
+			cxi->rh_files[j].midx[i] = rc;
+		}
+	}
+
+	/* List of per interface telemetry records */
+	rc = ldms_schema_metric_list_add(cxi->schema, "tel_list", "tel_record",
+					 ldms_record_heap_size_get(tel_rec) * cxi->iface_count);
+	cxi->tel_list_mid = rc;
+	if (rc < 0) {
+		ovis_log(cxi->log, OVIS_LERROR,
+			 "Error %d creating tel_list.\n", errno);
+		goto err;
+	}
+
+	/* List of per interface retry handler records */
+	rh_rec_mid = ldms_schema_record_add(cxi->schema, rh_rec);
+	rc = ldms_schema_metric_list_add(cxi->schema, "rh_list", "rh_record",
+					 ldms_record_heap_size_get(rh_rec) * cxi->iface_count);
+	cxi->rh_list_mid = rc;
+	if (rc < 0) {
+		ovis_log(cxi->log, OVIS_LERROR,
+			 "Error %d creating rh_list.\n", errno);
+		goto err;
+	}
+
+	size_t size = ldms_record_heap_size_get(tel_rec) +
+		      ldms_record_heap_size_get(rh_rec);
+	size = size * cxi->iface_count;
+	cxi->set = base_set_new_heap(cxi->base, size);
+	if (!cxi->set) {
+		rc = errno;
+		goto err;
+	}
+
+	ldms_mval_t rh_rec_mval;
+	ldms_mval_t tel_rec_mval;
+
+	for (i = 0; i < cxi->iface_count; i++) {
+		cxi->rh_list_mval = ldms_metric_get(cxi->set, cxi->rh_list_mid);
+		cxi->tel_list_mval = ldms_metric_get(cxi->set, cxi->tel_list_mid);
+		rh_rec_mval = ldms_record_alloc(cxi->set, rh_rec_mid);
+		tel_rec_mval = ldms_record_alloc(cxi->set, tel_rec_mid);
+
+		rc = ldms_list_append_record(cxi->set, cxi->rh_list_mval, rh_rec_mval);
+		if (rc) {
+			ovis_log(cxi->log, OVIS_LERROR,
+				 "Error %d appending record to rh_list.\n",
+				 rc);
+			goto err;
+		}
+
+		rc = ldms_list_append_record(cxi->set, cxi->tel_list_mval, tel_rec_mval);
+		if (rc) {
+			ovis_log(cxi->log, OVIS_LERROR,
+				 "Error %d appending record to tel_list.\n",
+				 rc);
+			goto err;
+		}
+	}
+	return 0;
+
+err:
+	if (cxi->base)
+		base_schema_delete(cxi->base);
```

Comment:
Should probably set ```cxi->base == NULL```  so we can't double call base_schema_delete() in destructor()

Reply:
Only if we set cxi->base to NULL after base_schema_delete(). Otherwise the pointer will be non-NULL, and the code in destructor() will try to access memory that has already been freed.